src/routing_table/routing_table.rs

```diff
@@ -0,0 +1,561 @@
+// Copyright 2015 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under (1) the MaidSafe.net Commercial License,
+// version 1.0 or later, or (2) The General Public License (GPL), version 3, depending on which
+// licence you accepted on initial access to the Software (the "Licences").
+//
+// By contributing code to the SAFE Network Software, or to this project generally, you agree to be
+// bound by the terms of the MaidSafe Contributor Agreement, version 1.0.  This, along with the
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use super::contact_info::ContactInfo;
+use super::result::{AddedNodeDetails, DroppedNodeDetails};
+use super::xorable::Xorable;
+
+use itertools::*;
+use std::{cmp, fmt, iter, mem, slice};
+
+type SliceFn<T> = fn(&Vec<T>) -> slice::Iter<T>;
+
+// Amount added to `min_bucket_len` when deciding whether a bucket split can happen.  This helps
+// protect against rapid splitting and merging in the face of moderate churn.
+const SPLIT_BUFFER: usize = 1;
+
+// Immutable iterator over the entries of a `RoutingTable`.
+pub struct Iter<'a, T: 'a> {
+    inner: iter::FlatMap<iter::Rev<slice::Iter<'a, Vec<T>>>, slice::Iter<'a, T>, SliceFn<T>>,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<&'a T> {
+        self.inner.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+
+// A message destination.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Destination<N> {
+    // The `k`-group of the given address. The message should reach the `k` closest nodes.
+    Group(N, usize),
+    // The individual node at the given address. The message should reach exactly one node.
+    Node(N),
+}
+
+impl<N> Destination<N> {
+    // Returns the name of the destination, i.e. the node or group address.
+    pub fn name(&self) -> &N {
+        match *self {
+            Destination::Group(ref name, _) |
+            Destination::Node(ref name) => name,
+        }
+    }
+
+    // Returns `true` if the destination os a group, and `false` if it is an individual node.
+    pub fn is_group(&self) -> bool {
```

Comment:
Here it may be worth adding `is_node' although just a negate of `is_group` it makes code easier to read.